CIP-hardforkSafeguard/README.md

```diff
@@ -0,0 +1,206 @@
+determine if the major protocol version will be increased at the next epoch
+boundary.
+The ledger itself does not apply the protocol parameter update until the
+epoch boundary.
+
+To apply the new safeguard, the consensus layer will now use new logic for
+determining if the major protocol will be increased, and the ledger will
+need to use the exact same logic on the epoch boundary.
+The new logic will take the same parameters that are currently being
+used to make the determination.
+See [protocolUpdates](https://github.com/input-output-hk/ouroboros-network/blob/314845c4087bc6e662d7df0d376ab1910a5b5476/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Inspect.hs#L106-L110),
+and note that the set of endorsements, the stake distribution, and the
+protocol parameters will all included in what the consensus layer calls the
+`LedgerState`, and what the ledger layer calls the `NewEpochState`
+(the endorsements will be added to `LedgerState`, but the pool stake
+distribution and the protocol parameters are already included).
+
+The new logic for determining if the major protocol version will change is:
+  * Has quorum been met on the proposed protocol parameter updates?
+    * If not, there is nothing else to do.
+    * If so, proceed.
+  * Does the update modify the major protocol parameter version?
+    * If not, the update will be applied on the epoch boundary, and there is
+      nothing else to do.
+    * If so, proceed.
+  * What is the sum of the relative, active stake of the block producers listed
+    in the endorsement set defined in the
+    [previous section](#tracking-hardfork-endorsements)?
+    Note that the stake distribution used here is the same as stake
+    distribution currently being used for block production.
+  * Is the sum computed above at least as large as the value of the
+    `hardforkThreshold` protocol parameter?
+    * If not, the entire update is rejected.
+    * If so, the update will be applied on the epoch boundary.
+
+#### Timing diagram
+
+The following table illustrates the timing described above,
+using the durations on mainnet (five-day epochs).
+
+```mermaid
+sequenceDiagram
+    participant s0 as Epoch Start
+    participant s1 as 12 hrs
+    participant s2 as 24 hrs
+    participant s3 as 36 hrs
+    participant s4 as 48 hrs
+    participant s5 as 60 hrs
+    participant s6 as 72 hrs
+    participant s7 as 84 hrs
+    participant s8 as 96 hrs
+    participant s9 as 108 hrs
+    participant sA as Epoch End
+
+    s0->s4: Proposal Window
+    s4->s7: Ledger state stabilization Window
+
+    Note over s7: Consensus to<br>determine if<br>hardfork will occur
+    Note over sA: Non-rejected<br>updates are applied<br>to ledger state
+```
+
+## Rationale
+
+The safeguard presented in this CIP aligns very closely with the manual check
+currently performed
+today before any hardfork.
+Moreover, we have strived to make the minimal changes needed to automate
```

Comment:
This is actually only part of the criteria currently being used. SPO block creation ratio, defi TVL criterion, exchange adoption criterion. I'm not saying that the other two should be encoded here - but it would be reasonable to mention them in the Rationale

Reply:
Maybe I can make it more clear that this CIP is only aiming to automate one very specific check? I myself do not know the whole process.

Reply:
Thanks for that clarification. Yes, that would be very appreciated. I clearly read way too much into what you were intending to change.